jt_notes.md

```diff
@@ -0,0 +1,127 @@
+
+------------------------------------ **Queries to Chas about the different modules ------------------------------------**
+
+cpi.py
+
+OpticalGater class that extends picamera.array.PiYUVAnalysis. It appears to maintain history, e.g. frame_history, period_history.
+
+**For chas:** my opinion is frame_history is really crying out for a class container, or at least an array with named columns (or a list-of-dictionaries, though I'm less keen on that). At present it requires the developer to keep track of the meaning of each column, and the format not easily extensible.
+
+**For chas:** load_data() - "Place holder function for loading data in emulator". What is the intention of this, and is there a clear idea of what an emulator would look like?
+
+**For me:** this is a general question for Chas: how does he intend the emulator to work? I don't think I've fully followed his emulator stuff in "app". Might need to talk through that one together...
+
+**For chas: IMPORTANT, BUG:** as per my comment in this module, I don't think barrier frames are implemented properly.
+
+**For chas: IMPORTANT, BUG:** I think there is a bug associated with (lack of) numExtraRefFrames in initialise_internal_parameters(), as per my comment there.
+
+**For chas: IMPORTANT, approaching a BUG**: there is the general issue that the code is not doing a great job of precise timing. It determines a delay time before sending the trigger, but then executes a bunch more code. Oh, and that delay time is then treated relative to "current_time", which is set *after* doing the phase-matching. This is going to reduce accuracy and precision, and also makes me even more uncomfortable in terms of future-proofing. I think it would be much better to pass around absolute times, not deltas.
+
+**For chas**: more generally, I wish we were working with frame *objects* that have accompanying metadata (such as timestamp!). That would help solve the above issue about timings in a tidy manner, and also make it a lot easier to generalise to other timebases (such as camera timebase, rather than computer). Oh, and it would also help deal with gaps in the timebase (missing frames), which I think would be a useful thing to cope with, even if right now there might not be any concrete scenario in which frames would be dropped. I fear I may not win that argument about frame objects, since it would be quite a big structural change. But, do you have any alternative ideas about how best to cope with the fact that we may well want to work in terms of timebases other than the computer's own timebase?
```

Comment:
I want there to be a nice numpy solution to this but I'm not sure there is. One thin that jumps to mind is using hdf5 because you can give arrays custom attributes - but that's really not worth if we're keeping the frames off disk.... however.... as we do tend to save our bf files to disk.... hdf5 could kill to birds with one lovely stone?